templates/terraform/resource_iam.html.markdown.erb

```diff
@@ -0,0 +1,156 @@
+    fine - but when you need to add custom docs (notes, etc), you need
+    to remember these things.
+
+    Know also that the `lines` function in heavy use in MagicModules will
+    strip exactly one trailing newline - unless that's what you've designed
+    your docstring for, it's easier to insert newlines where you need them
+    manually.  That's why, in this file, we use `lines` on anything which
+    is generated from a ruby function, but skip it on anything that is
+    directly inserted from YAML.
+-%>
+<%
+  if @config.legacy_name.nil?
+    terraform_name = "google_" + (product_ns + object.name).underscore + "_iam"
+  else
+    terraform_name = "google_" + @config.legacy_name + '_' + object.name.underscore + "_iam"
+  end
+  properties = object.all_user_properties
+  timeouts = object.async.nil? ? Api::Timeouts.new : object.timeouts
+-%>
+---
+<%= lines(autogen_notice :yaml) -%>
+layout: "google"
+page_title: "Google: <%= terraform_name -%>"
+sidebar_current: "docs-<%= terraform_name.gsub("_", "-") -%>"
+description: |-
+  Collection of resources to manage IAM policy for <%= product_ns -%> <%= object.name -%>
+
+Three different resources help you manage your IAM policy for <%= product_ns -%> <%= object.name -%>. Each of these resources serves a different use case:
+
+* `<%= terraform_name -%>_policy`: Authoritative. Sets the IAM policy for the <%= object.name.downcase -%> and replaces any existing policy already attached.
+* `<%= terraform_name -%>_binding`: Authoritative for a given role. Updates the IAM policy to grant a role to a list of members. Other roles within the IAM policy for the <%= object.name.downcase -%> are preserved.
+* `<%= terraform_name -%>_member`: Non-authoritative. Updates the IAM policy to grant a role to a new member. Other members for the role for the <%= object.name.downcase -%> are preserved.
+
+~> **Note:** `<%= terraform_name -%>_policy` **cannot** be used in conjunction with `<%= terraform_name -%>_binding` and `<%= terraform_name -%>_member` or they will fight over what your policy should be.
+
+~> **Note:** `<%= terraform_name -%>_binding` resources **can be** used in conjunction with `<%= terraform_name -%>_member` resources **only if** they do not grant privilege to the same role.
+
+<% if object.min_version.name == 'beta' -%>
+~> **Warning:** This resource is in beta, and should be used with the terraform-provider-google-beta provider.
+See [Provider Versions](https://terraform.io/docs/providers/google/provider_versions.html) for more details on beta resources.
+<% end -%>
+
+<% markdown_escaped_name = terraform_name.gsub("_", "\\_") %>
+## <%= markdown_escaped_name -%>\_policy
+
+```hcl
+data "google_iam_policy" "admin" {
+  binding {
+    role    = "roles/editor"
+    members = [
+      "user:jane@example.com",
+    ]
+  }
+}
+
+resource "<%= terraform_name -%>_policy" "editor" {
+  <%= object.name.underscore -%>       = "<%= object.id_format.gsub('{{name}}', "{{#{object.name.underscore}}}")-%>"
```

Comment:
Is the change to partial URI in pubsub intentional? Will specifying that work? https://github.com/terraform-providers/terraform-provider-google-beta/pull/914/files#diff-5bc4b7188e596ed679a95ddfd0c495c2R45

Reply:
Yeah, I believe changing to the partial URI makes the tests and examples more deterministic. Currently the pubsub topic IAM resources allow many different ways of specifying the topic that is the subject of IAM policies. For example `topic = "topic-name"` uses the default project, but still works. Tests of these different ways of declaring it can be found here: https://github.com/terraform-providers/terraform-provider-google-beta/blob/master/google-beta/resource_pubsub_topic_iam_test.go and are still being used to test the generated version

It makes sense to me to be as specific as possible in test/example code rather than leaving the project unspecified.  It also allows for the generation to be more generic

Reply:
I'd guess that most users expect to use short names, and so I lean slightly more towards short names than partial URIs. That said, not strongly enough to block- the distinction doesn't end up mattering much, and it's easier for us if users specify the partial URI as well.